### ISSUE TITLE
EditorView & UITextViewDelegate

### ISSUE THREAD

**Issue by writing-shed** [rajdeep/proton#85]
I am trying to add Proton to a SwiftUI app I've been developing for about a year. It contains a UITextView editor embedded in a UIViewRepresentable class. I want to have bullet/numbered lists and embedded images, which is not easy as I'm sure you know. In fact, trying to solve the problem is how I found Proton. Congratulations on a great piece of work!

I have experimented with the ExampleApp to get a feel for things and have managed to embed a Proton EditorView in my app as a replacement for the UITextView. I can type into it, but that's about all. What I am missing is the Proton equivalents for UITextViewDelegate methods. I am also struggling to detect the EditorView's firstResponder state.

Finally, as you know, the Master package doesn't compile because of the mixed language issue, so I am using version 0.5 at the moment.

**Comment by rajdeep** [comment]
Hi @writing-shed,
Glad that you like Proton.  I do not have SwiftUI support/examples as yet but that is something which is in my near term roadmap. However, given the state of SwiftUI and the fact that Proton doesn't have any support as yet, it might take some time. If you are interested in contributing ideas, suggestions or even issues that you have faced, please feel free to join the discussion: #83.

In regards to `UITextViewDelegate`, the equivalent would be `EditorViewDelegate`. It provides certain specific callbacks that might be of help to you. You can get to know about the firstResponder state using `editor(_:didReceiveFocusAt:)` and `editor(_:didLoseFocusFrom:)`. You can find the documentation here: https://rajdeep.github.io/proton/Protocols/EditorViewDelegate.html. If there's something that you find missing, please feel free to add to this issue.

**Comment by writing-shed** [comment]
Hi Rajdeep

Thanks for your reply. I had already tried using EditorViewDelegate but stupidly got the func signatures wrong! I've corrected them and it seems to be working, though I've still got some way to go to properly consume Proton.

I see that #83 was started by Vatsal Manot - I have been using parts of his SwiftUIX for over six months now. I will contribute to the thread when/if I think I have something useful to say. At the moment it looks like Proton is a plug in replacement for UITextView which has no SwiftUI equivalent. SwiftUI has recently been blessed with a TextEditor struct, This is pitifully weak  in comparison to UITextView. I wrote my first app for writers just after Mac Classic appeared complete with MacWrite. Then, as now, Apple bury useful text editing features in the OS. A glance at Pages will show those features not available to developers (not least bullet lists, numbered lists, embedded graphics). TextKit on IOS is not the same as TextKit on Mac OS - LayoutManager in IOS doesn't have a list of UITextViews, ergo a UITextView cannot have its own LayoutManager - see LayoutManager.firstTextView). Anyway, I'm rambling. I will look out for Proton features that could benefit from SwiftUI support.

**Comment by rajdeep** [comment]
I concur with your thoughts, @writing-shed. Apple certainly has kept a lot of useful features from developers. I hope with time, Proton will be able to help solve some of these issues and have usable SwiftUI support.

**Comment by writing-shed** [comment]
When the Mac Classic came out I became a registered developer (1984 I think). The only documentation was collections of API definitions that could be downloaded from the Apple network (this was pre-web). One of these defined the API that was obviously used by MacWrite, so I started to use it. At the time all development was in Pascal, cross compiled and tested from a Lisa. I pretty soon discovered that the API I was using wasn’t actually visible to developers!! Not a lot has changed. It bugs me that they haven’t even published the Pages document format.

Anyway, less of me carping about ancient history. As I said I am well into developing a replacement for Writing Shed, my current app, and so all my energy is going there. I will, however, try and create a SwiftUI interface for EditorView.

Best
Keith

BTW I am in the UK 11 hours behind you.

**Comment by vmanot** [comment]
@writing-shed I might be able to save you some work on the SwiftUI port, since I'm working on a similar project. Let me know if you're interested in collaborating.

**Comment by writing-shed** [comment]
Thanks for the offer. I’m totally committed to completing my app at the moment and have only made a small change in direction to embed Proton. I am sure you will make a great job of porting it to SwiftUI. I would be quite happy to help you test it once I’ve got past my list of to do items, possibly embedding it in my app.

Best
Keith

**Comment by niklasgrewe** [comment]
Hi @writing-shed @vmanot  
i just came across Proton by accident. I'm also currently developing an iOS app in SwiftUI and am looking for a native flexible rich text editor. It has been some time since January. Can you tell me any experiences? How well does Proton work in SwiftUI? Am looking forward to your reply.

Best regards
Niklas

**Comment by writing-shed** [comment]
Hi 

Yes, I am still using Proton and it is working very well for me. I still have it sitting inside UIViewRepresentable. I have had a few hiccups, all down to the way I was handling the data and not understanding Lists properly. TBH the fact that I am using SwiftUI  has not impacted on the way I use Proton. It would be pretty much the same if I was using UIKIt. The big plus point is that I am building a multi-platform app with relative ease and Proton is quite happy with Mac OS. I was slightly concerned when I learned about TextKit 2 in WWDC 21. Looking more closely I noticed that Apple have not provided any replacement for UITextView to enable it to be used by an editor. I'm sure it will come, possibly next June. 

Regards
Keith
